### PR TITLE
🐛 fix: allow reconciliation of deadline-exceeded ClusterObjectSets

### PR DESCRIPTION
**Pull request by joelanford**
# Description

The `skipProgressDeadlineExceededPredicate` blocked all update events for COS objects
with `ProgressDeadlineExceeded`, which prevented archival of stuck revisions — the
lifecycle state patch was silently dropped.

This PR:
- **Removes the predicate** so all COS events are fully reconciled
- **Updates `markAsProgressing`** to set `ProgressDeadlineExceeded` instead of
  `RollingOut`/`Retrying` when the deadline has been exceeded, preventing the reconcile
  loop the predicate was masking
- **Continues reconciling after `ProgressDeadlineExceeded`** rather than clearing the
  error and stopping requeue. This allows revisions to recover if a transient error
  resolves itself, even after the deadline was exceeded
- **Extracts `durationUntilDeadline`** as a shared helper for deadline computation
- **Adds a `deadlineAwareRateLimiter`** that caps exponential backoff at the deadline so
  `ProgressDeadlineExceeded` is set promptly even during error retries
- **Moves deadline requeue logic** into `reconcile`, using the remaining duration from
  `durationUntilDeadline` directly for `RequeueAfter`
- **Adds unit tests** for `durationUntilDeadline` and recovery from
  `ProgressDeadlineExceeded` to `Succeeded`
- **Adds e2e tests** for `ProgressDeadlineExceeded` archival/cleanup and for recovery
  from `ProgressDeadlineExceeded` to `Succeeded`

Addresses feedback from:
https://github.com/operator-framework/operator-controller/pull/2610#issuecomment-4226236991

## Reviewer Checklist

- [x] Tests: Unit Tests (and E2E Tests, if appropriate)
- [x] Comprehensive Commit Messages
- [x] Links to related GitHub Issue(s)

### CODE REVIEW

File: test/e2e/features/revision.feature [hunk 1]
```diff
@@ -442,4 +442,71 @@ Feature: Install ClusterObjectSet
...
+                      image: does-not-exist:latest
+        revision: 1
+      """
+    Then resource "configmap/test-configmap" is installed
+    And resource "deployment/test-deployment" is installed
+    And ClusterObjectSet "${COS_NAME}" reports Progressing as False with Reason ProgressDeadlineExceeded
+    When ClusterObjectSet "${COS_NAME}" lifecycle is set to "Archived"
+    Then ClusterObjectSet "${COS_NAME}" is archived
+    And resource "configmap/test-configmap" is eventually not found
+    And resource "deployment/test-deployment" is eventually not found
```
Comment (camilamacedo86) [inline test/e2e/features/revision.feature:512]: It does not seems to be testing the same scenario @joelanford 
Could we ensure the same scenario here?

Comment (camilamacedo86) [reply test/e2e/features/revision.feature:512]: 1. User installs a ClusterExtension. The CE controller creates COS-rev-1.
2. COS-rev-1 gets stuck (e.g. a Deployment never becomes ready). After ProgressDeadlineMinutes, the reconciler sets Progressing=False/ProgressDeadlineExceeded.
4. User updates the ClusterExtension. The CE controller creates COS-rev-2.
5. COS-rev-2 rolls out successfully. It patches COS-rev-1 with lifecycleState: Archived so the old revision gets cleaned up.
7. The watch predicate sees COS-rev-1 has ProgressDeadlineExceeded and drops the event.
8. COS-rev-1 never reconciles, never processes the archival, and stays stuck forever.

Comment (joelanford) [reply test/e2e/features/revision.feature:512]: Is it not the same? Making a CE that stamps out the COS-1 and COS-2 such that the CE reconciler eventually tries to set COS-1 as archived is the same thing, but just more ceremony around a standalone COS that exceeds the deadline and then is directly archived by the test code.

I'll verify manually to make sure. Is there something else happening with the CE/COS interaction that would make the COS-only test in the PR different?

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 2]
```diff
@@ -367,6 +346,51 @@ func (c *ClusterObjectSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	clk clock.Clock,
+) *deadlineAwareRateLimiter {
+	return &deadlineAwareRateLimiter{delegate: delegate, client: c, clock: clk}
+}
+
+func (r *deadlineAwareRateLimiter) When(item ctrl.Request) time.Duration {
+	backoff := r.delegate.When(item)
+
+	cos := &ocv1.ClusterObjectSet{}
+	if err := r.client.Get(context.Background(), item.NamespacedName, cos); err != nil {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:371]: Our use of the rate limiter uses the informer cache which is in-memory and does not block.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 3]
```diff
@@ -333,29 +323,19 @@ type Sourcoser interface {
...
-			if !rev.DeletionTimestamp.IsZero() {
-				return true
-			}
-			if cnd := meta.FindStatusCondition(rev.Status.Conditions, ocv1.ClusterObjectSetTypeProgressing); cnd != nil && cnd.Status == metav1.ConditionFalse && cnd.Reason == ocv1.ReasonProgressDeadlineExceeded {
-				return false
-			}
-			return true
-		},
-	}
 	c.Clock = clock.RealClock{}
```
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:326]: Not something I changed in this PR, so out-of-scope. Regardless, caller's don't call SetupWithManager in tests, so this isn't a problem in practice.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 4]
```diff
@@ -367,6 +347,51 @@ func (c *ClusterObjectSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	clk clock.Clock,
+) *deadlineAwareRateLimiter {
+	return &deadlineAwareRateLimiter{delegate: delegate, client: c, clock: clk}
+}
+
+func (r *deadlineAwareRateLimiter) When(item ctrl.Request) time.Duration {
+	backoff := r.delegate.When(item)
+
+	cos := &ocv1.ClusterObjectSet{}
+	if err := r.client.Get(context.Background(), item.NamespacedName, cos); err != nil {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:371]: Our use of the rate limiter uses the informer cache which is in-memory and does not block.

File: test/e2e/features/revision.feature (hunk 1 above)
Comment (camilamacedo86) [reply test/e2e/features/revision.feature:512]: we are not upgrading, checking it stuck and then, checking that is possible to change the resource and the changes will be reconciled. WDY about we add a test with?

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 5]
```diff
@@ -144,15 +122,18 @@ func (c *ClusterObjectSetReconciler) reconcile(ctx context.Context, cos *ocv1.Cl
 		return c.delete(ctx, cos)
 	}
 
+	remaining, hasDeadline := c.durationUntilDeadline(cos)
+	isDeadlineExceeded := hasDeadline && remaining <= 0
+
```
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:127]: This is a good point. `requeueForDeadline` should not perform the calculation again with a different "now" timestamp. Perhaps instead we could simply pass the result of the original calculation into requeueForDeadline.

**Comment by camilamacedo86** [comment]
@joelanford can you please rebase this one and address the Copilot comments so that we can review and see if we can go with.

**Comment by joelanford** [comment]
Ran out of time to get back to this today. Hopefully tomorrow!

This PR has some flaws around recomputation of the remaining time until deadline which can cause funky things to happen if we're reconciling just before the deadline and then cross over the deadline threshold prior to the rate limiter logic running, which can cause requeues not to happen right at the deadline.

Might need a bit more logic in the recomputation code before it's ready. Hopefully that extra code isn't too brittle. Maybe it will be fine as a stopgap if the plan is to eventually move the progress deadline stuff to be a concern only of the ClusterExtension (leaving ClusterObjectSet with just a single "Ready" or similar condition)

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 6]
```diff
@@ -288,9 +267,7 @@ func (c *ClusterObjectSetReconciler) reconcile(ctx context.Context, cos *ocv1.Cl
 		} else {
 			markAsUnavailable(cos, ocv1.ReasonRollingOut, fmt.Sprintf("Revision %s is rolling out.", revVersion))
 		}
-		if meta.FindStatusCondition(cos.Status.Conditions, ocv1.ClusterObjectSetTypeProgressing) == nil {
-			markAsProgressing(cos, ocv1.ReasonRollingOut, fmt.Sprintf("Revision %s is rolling out.", revVersion))
-		}
+		markAsProgressing(cos, ocv1.ReasonRollingOut, fmt.Sprintf("Revision %s is rolling out.", revVersion))
```
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go]: Ah right. I had fixed this in my original attempt, but reverted because I had forgotten about the flapping that a post-reconcile fixup was causing.

I put back the logic to set this correctly at the markAsProgressing call, with no need for a post-reconcile fixup.

**Comment by joelanford** [comment]
~experimental e2e is timing out. I guess I need to bump e2e timeout again?~

Never mind: just needed to pick up the latest changes from `main`

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 7]
```diff
@@ -659,14 +631,26 @@ func buildProgressionProbes(progressionProbes []ocv1.ProgressionProbe) (probing.
...
+	case ocv1.ReasonSucceeded:
+		// Terminal — always apply.
+	case ocv1.ReasonRollingOut, ocv1.ClusterObjectSetReasonRetrying:
+		if isDeadlineExceeded {
+			markAsNotProgressing(cos, ocv1.ReasonProgressDeadlineExceeded,
+				fmt.Sprintf("Revision has not rolled out for %d minute(s). Last status: %s", cos.Spec.ProgressDeadlineMinutes, message))
+			return
+		}
+	default:
+		panic(fmt.Sprintf("unregistered progressing reason: %q", reason))
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go]: is panicking the right strategy here... it is going to restart the pod, but with the next reconcile of the problematic resource, it is going to crash again affecting reconciliation of other valid resources.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 8]
```diff
@@ -659,14 +631,26 @@ func buildProgressionProbes(progressionProbes []ocv1.ProgressionProbe) (probing.
...
 	if meta.FindStatusCondition(cos.Status.Conditions, ocv1.ClusterObjectSetTypeAvailable) != nil {
 		markAsAvailableUnknown(cos, ocv1.ClusterObjectSetReasonReconciling, message)
 	}
 }
 
-func markAsProgressing(cos *ocv1.ClusterObjectSet, reason, message string) {
+func markAsProgressing(cos *ocv1.ClusterObjectSet, reason, message string, isDeadlineExceeded bool) {
+	switch reason {
+	case ocv1.ReasonSucceeded:
+		// Terminal — always apply.
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go:652]: **No test for the recovery-after-ProgressDeadlineExceeded path.**

The `ReasonSucceeded` case is marked "Terminal — always apply", which means a COS that hits `ProgressDeadlineExceeded` and later has its probes pass should transition back to `Succeeded`. The PR description claims this behavior: _"allows revisions to recover if a transient error resolves itself, even after the deadline was exceeded."_

But neither the unit tests nor the e2e test exercise this recovery path. A unit test showing a COS transitioning from `ProgressDeadlineExceeded` → `Succeeded` when probes start passing would document and protect this behavior.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 9]
```diff
@@ -146,6 +124,9 @@ func (c *ClusterObjectSetReconciler) reconcile(ctx context.Context, cos *ocv1.Cl
 		return c.delete(ctx, cos)
 	}
 
+	remaining, hasDeadline := durationUntilDeadline(c.Clock, cos)
+	isDeadlineExceeded := hasDeadline && remaining <= 0
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go:129]: **Behavioral change: errors are no longer swallowed after deadline.**

The old code explicitly cleared errors when the deadline was exceeded:
```go
reconcileErr = nil
res = ctrl.Result{}
```
The new code propagates errors even after the deadline, relying on the sticky `ProgressDeadlineExceeded` condition to communicate state. This is intentional per the PR description, but it changes how the controller interacts with the workqueue — the COS will keep retrying with backoff rather than going quiet.

Worth confirming this doesn't cause excessive log noise or API server load for permanently broken COS objects that will never recover.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 10]
```diff
@@ -659,14 +631,26 @@ func buildProgressionProbes(progressionProbes []ocv1.ProgressionProbe) (probing.
...
 
-func markAsProgressing(cos *ocv1.ClusterObjectSet, reason, message string) {
+func markAsProgressing(cos *ocv1.ClusterObjectSet, reason, message string, isDeadlineExceeded bool) {
+	switch reason {
+	case ocv1.ReasonSucceeded:
+		// Terminal — always apply.
+	case ocv1.ReasonRollingOut, ocv1.ClusterObjectSetReasonRetrying:
+		if isDeadlineExceeded {
+			markAsNotProgressing(cos, ocv1.ReasonProgressDeadlineExceeded,
+				fmt.Sprintf("Revision has not rolled out for %d minute(s). Last status: %s", cos.Spec.ProgressDeadlineMinutes, message))
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go:659]: **Condition message can grow large.**

The inner `message` can already be quite verbose — e.g., probe failure messages include full `Kind.Group Namespace/Name` strings, or multi-line collision reports from `strings.Join(collidingObjs, "\n\n")`. Concatenating it untruncated into `"Last status: %s"` could produce very long status condition messages. Consider truncating `message` before embedding, or capping the total length.

File: internal/operator-controller/controllers/progress_deadline.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+// alone that a COS succeeded in the past. If Succeeded is removed or set to
+// False, this function will return a deadline and the reconciler will set
+// ProgressDeadlineExceeded even though the revision previously succeeded.
+//
+// Returns (0, false) when there is no active deadline:
+//   - progressDeadlineMinutes is 0
+//   - the revision has already succeeded
+//   - the revision is archived (deadline is irrelevant)
+//   - the revision is being deleted
+func durationUntilDeadline(clk clock.Clock, cos *ocv1.ClusterObjectSet) (time.Duration, bool) {
```
Comment (pedjak) [inline internal/operator-controller/controllers/progress_deadline.go:94]: **`durationUntilDeadline` has no direct unit tests.**

This function has explicit early-return branches for `DeletionTimestamp != zero`, `LifecycleState == Archived`, `Succeeded == True`, and `pd <= 0`, but `progress_deadline_test.go` only tests the rate limiter (which calls `durationUntilDeadline` indirectly). Adding targeted table-driven tests for each early-return branch would protect against regressions if this logic changes — especially since the PR description documents the Succeeded-latch semantics as an important design decision.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 12]
```diff
@@ -146,6 +124,9 @@ func (c *ClusterObjectSetReconciler) reconcile(ctx context.Context, cos *ocv1.Cl
 		return c.delete(ctx, cos)
 	}
 
+	remaining, hasDeadline := durationUntilDeadline(c.Clock, cos)
+	isDeadlineExceeded := hasDeadline && remaining <= 0
+
 	if err := c.verifyReferencedSecretsImmutable(ctx, cos); err != nil {
 		l.Error(err, "referenced Secret verification failed, blocking reconciliation")
 		markAsNotProgressing(cos, ocv1.ClusterObjectSetReasonBlocked, err.Error())
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go:134]: **`verifyReferencedSecretsImmutable` bypasses deadline-aware conditions.**

When secret verification fails, this calls `markAsNotProgressing(cos, Blocked, ...)` directly, not through `setRetryingConditions`. This means a COS that is both blocked AND past its deadline shows `Blocked` rather than `ProgressDeadlineExceeded`.

This is probably the right priority ordering (Blocked is more actionable), but it's an implicit precedence that isn't documented or tested. A brief comment here, or a test case covering the blocked+deadline-exceeded scenario, would clarify the intent.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 13]
```diff
@@ -146,6 +124,9 @@ func (c *ClusterObjectSetReconciler) reconcile(ctx context.Context, cos *ocv1.Cl
 		return c.delete(ctx, cos)
 	}
 
+	remaining, hasDeadline := durationUntilDeadline(c.Clock, cos)
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go:128]: **`isDeadlineExceeded` is computed once, used after potentially long operations.**

`durationUntilDeadline` is called here at the start of `reconcile()`, but the resulting `isDeadlineExceeded` flag is threaded through calls that happen after `buildBoxcutterPhases`, `establishWatch`, and `revisionEngine.Reconcile`. If those operations take significant time (especially `Reconcile`), the snapshot could be stale — a COS might cross the deadline mid-reconcile but still get `RollingOut` instead of `ProgressDeadlineExceeded`.

In practice the rate limiter handles the next iteration, and recomputing mid-reconcile risks the "different Now" problem already discussed. Worth acknowledging as an accepted trade-off.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 14]
```diff
@@ -304,8 +285,9 @@ func (c *ClusterObjectSetReconciler) reconcile(ctx context.Context, cos *ocv1.Cl
 		} else {
 			markAsUnavailable(cos, ocv1.ReasonRollingOut, fmt.Sprintf("Revision %s is rolling out.", revVersion))
 		}
-		if meta.FindStatusCondition(cos.Status.Conditions, ocv1.ClusterObjectSetTypeProgressing) == nil {
-			markAsProgressing(cos, ocv1.ReasonRollingOut, fmt.Sprintf("Revision %s is rolling out.", revVersion))
+		markAsProgressing(cos, ocv1.ReasonRollingOut, fmt.Sprintf("Revision %s is rolling out.", revVersion), isDeadlineExceeded)
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go]: **Removed the "set progressing only if nil" guard.**

The old code had:
```go
if meta.FindStatusCondition(cos.Status.Conditions, ocv1.ClusterObjectSetTypeProgressing) == nil {
    markAsProgressing(cos, ocv1.ReasonRollingOut, ...)
}
```
This now unconditionally calls `markAsProgressing`, which means every reconcile where probes are failing will overwrite the Progressing condition. If the COS already has `ProgressDeadlineExceeded` (Progressing=False), this will re-set it via `markAsNotProgressing` inside `markAsProgressing`.

The behavior is correct because `meta.SetStatusCondition` only updates `LastTransitionTime` when status or reason actually changes — so this is idempotent. But it's a subtlety worth a brief comment, since the old guard existed for a reason and a future reader might wonder why it was removed.

**Review by pedjak — COMMENTED** [review]
Additional review comments — focused on test coverage gaps, behavioral subtleties, and edge cases.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 7 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go]: Idea is that if we add a new reason, it needs to be explicitly mentioned here to decide if it is a "success" reason or a "still progressing" reason so that it gets the correct deadline exceeded override treatment.

Seemed better to me to fail loud and hard so that we catch it in development.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 9 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:129]: That is intentional. Exponential backoff after the deadline will avoid excessive log noise and API server load. And it is a fallacy to assume that a COS that has exceeded it's deadline is permanently broken. They very well may recover after the deadline.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 10 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:659]: This logic merely adds a fixed amount of length to the message that was passed into the function. Before, the intended message was swallowed. Now the underlying issue causing the errors is still surfaced. It might be essential for a user to be able to troubleshoot the issue.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 7 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go]: Fair point. Changed this to log an error and treat the unknown reason as non-terminal (i.e., it will still flip to ProgressDeadlineExceeded when the deadline is exceeded). This avoids crash-looping the pod while still making the issue visible in logs.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 8 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:652]: Added a unit test: "recovery from ProgressDeadlineExceeded to Succeeded when revision completes". It sets up a COS with Progressing=False/ProgressDeadlineExceeded, returns a complete revision result, and verifies the condition transitions to Progressing=True/Succeeded.

File: internal/operator-controller/controllers/progress_deadline.go (hunk 11 above)
Comment (joelanford) [reply internal/operator-controller/controllers/progress_deadline.go:94]: Added table-driven tests covering each early-return branch: pd<=0, Succeeded=True, Archived, DeletionTimestamp set, deadline not yet exceeded, and deadline already exceeded.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 12 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:134]: Added a comment clarifying the intent: "Blocked takes precedence over ProgressDeadlineExceeded: it is more actionable for the user."

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 13 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go:128]: Agreed this is an accepted trade-off. The rate limiter handles the next iteration promptly, and recomputing mid-reconcile would introduce the "different Now" problem discussed in the earlier Copilot thread.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 14 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go]: This is safe because `meta.SetStatusCondition` only updates `LastTransitionTime` when status or reason actually changes, making the call idempotent. The old guard prevented deadline enforcement from working correctly on subsequent reconciles. It is also an anti-pattern to take action based on a status read of the object we are reconciling.

File: internal/operator-controller/controllers/clusterobjectset_controller_test.go [hunk 15]
```diff
@@ -1000,6 +1000,39 @@ func Test_ClusterObjectSetReconciler_Reconcile_ProgressDeadline(t *testing.T) {
 				require.Equal(t, ocv1.ReasonRollingOut, cnd.Reason)
 			},
 		},
+		{
+			name: "recovery from ProgressDeadlineExceeded to Succeeded when revision completes",
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller_test.go:1004]: **Consider an e2e test for the recovery path.**

This unit test mocks away the workqueue, rate limiter, watches, and event pipeline — the exact components this PR changes. It proves the `markAsProgressing` switch logic works, but doesn't validate that a deadline-exceeded COS will actually get re-reconciled in a running controller.

A simple e2e test could cover this without any mid-test patching: use a deployment with a delayed readiness probe (`initialDelaySeconds: 80`) and `progressDeadlineMinutes: 1`. The COS hits `ProgressDeadlineExceeded` at ~60s, the deployment becomes ready at ~80s, and the COS should recover to `Succeeded` — testing the full loop (workqueue re-enqueue, rate limiter, reconcile, probe pass, condition flip).

```gherkin
Then ClusterObjectSet "${COS_NAME}" reports Progressing as False with Reason ProgressDeadlineExceeded
Then ClusterObjectSet "${COS_NAME}" reports Progressing as True with Reason Succeeded
```

Adds ~90 seconds to the suite, roughly the same as the archival scenario already in this PR.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 16]
```diff
@@ -659,14 +632,32 @@ func buildProgressionProbes(progressionProbes []ocv1.ProgressionProbe) (probing.
...
+func setRetryingConditions(cos *ocv1.ClusterObjectSet, message string, isDeadlineExceeded bool) {
+	markAsProgressing(cos, ocv1.ClusterObjectSetReasonRetrying, message, isDeadlineExceeded)
 	if meta.FindStatusCondition(cos.Status.Conditions, ocv1.ClusterObjectSetTypeAvailable) != nil {
 		markAsAvailableUnknown(cos, ocv1.ClusterObjectSetReasonReconciling, message)
 	}
 }
 
-func markAsProgressing(cos *ocv1.ClusterObjectSet, reason, message string) {
+func markAsProgressing(cos *ocv1.ClusterObjectSet, reason, message string, isDeadlineExceeded bool) {
+	nonTerminalReasons := map[string]struct{}{
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go]: Nit: this map is allocated on every `markAsProgressing` call (multiple times per reconcile). Could be a package-level `var` for a free optimization.

File: internal/operator-controller/controllers/clusterobjectset_controller.go [hunk 17]
```diff
@@ -659,14 +632,32 @@ func buildProgressionProbes(progressionProbes []ocv1.ProgressionProbe) (probing.
...
+		ocv1.ReasonRollingOut:               {},
+		ocv1.ClusterObjectSetReasonRetrying: {},
+	}
+
+	switch reason {
+	case ocv1.ReasonSucceeded:
+		// Terminal — always apply.
+	default:
+		if _, known := nonTerminalReasons[reason]; !known {
+			log.Log.Error(fmt.Errorf("unregistered progressing reason: %q", reason), "treating as non-terminal for deadline enforcement")
```
Comment (pedjak) [inline internal/operator-controller/controllers/clusterobjectset_controller.go]: Nit: `log.Log` is the global logger — the error message won't include structured context (e.g., COS name), making it harder to diagnose in production. Acceptable given the function signature lacks a `context.Context`, but worth noting.

**Review by pedjak — COMMENTED** [review]
Follow-up comments on the updated PR — one test coverage suggestion and two minor nits.

File: internal/operator-controller/controllers/clusterobjectset_controller_test.go (hunk 15 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller_test.go:1004]: Added.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 17 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go]: Plumbed the context's logger.

File: internal/operator-controller/controllers/clusterobjectset_controller.go (hunk 16 above)
Comment (joelanford) [reply internal/operator-controller/controllers/clusterobjectset_controller.go]: Done

**Review by pedjak — COMMENTED** [review]
/approve

**Review by camilamacedo86 — APPROVED** [review]
